share/man/man8/automount.8
.\"
.\" The contents of this file are subject to the terms of the
.\" Common Development and Distribution License (the "License").
.\" You may not use this file except in compliance with the License.
.\"
.\" You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
.\" or http://www.opensolaris.org/os/licensing.
.\" See the License for the specific language governing permissions
.\" and limitations under the License.
.\"
.\" When distributing Covered Code, include this CDDL HEADER in each
.\" file and include the License file at usr/src/OPENSOLARIS.LICENSE.
.\" If applicable, add the following below this CDDL HEADER, with the
.\" fields enclosed by brackets "[]" replaced with your own identifying
.\" information: Portions Copyright [yyyy] [name of copyright owner]
.\"
.\"
.\" Copyright 1989 AT&T
.\" Copyright (c) 2008, Sun Microsystems, Inc. All Rights Reserved
.\" Copyright 2016 Nexenta Systems, Inc.
.\"
.Dd July 16, 2018
.Dt AUTOMOUNT 8
.Os
.Sh NAME
.Nm automount
.Nd install automatic mount points
.Sh SYNOPSIS
.Nm
.Op Fl v
.Op Fl t Ar duration
.Sh DESCRIPTION
The
.Nm
utility installs
.Nm autofs
mount points and associates an automount map with each mount point. It starts
the
.Xr automountd 8
daemon if it finds any non-trivial entries in either local or distributed
automount maps and if the daemon is not already running. The
.Nm autofs
file system monitors attempts to access directories within it and notifies the
.Xr automountd 8
daemon. The daemon uses the map to locate a file system, which it then mounts at
the point of reference within the
.Nm autofs
file system. A map can be assigned to an
.Nm autofs
mount using an entry in the
.Pa /etc/auto_master
map or a direct map.
.Pp
If the file system is not accessed within an appropriate interval
.Pq 10 minutes by default ,
the
.Nm automountd
daemon unmounts the file system.
.Pp
The file
.Pa /etc/auto_master
determines the locations of all
.Nm autofs
mount points.
.Pp
The
.Nm
command is usually run without arguments. It compares the entries
.Pa /etc/auto_master
with the current list of
.Nm autofs
mounts in
.Pa /etc/mnttab
and adds, removes or updates
.Nm autofs
mounts to bring the
.Pa /etc/mnttab
up to date with the
.Pa /etc/auto_master .
At boot time it installs all
.Nm autofs
mounts from the master map. Subsequently, it may be run to install
.Nm autofs
mounts for new entries in the master map or the direct map, or to perform
unmounts for entries that have been removed from these maps.
.Sh OPTIONS
The following options are supported:
.Bl -tag -width Ds
.It Fl v
Verbose mode. Notifies of
.Nm autofs
mounts, unmounts, or other non-essential information.
.It Fl t Ar duration
Specifies a
.Ar duration ,
in seconds, that a file system is to remain mounted when not in use. The default
is
.Sy 10
minutes.
.El
.Sh USAGE
.Ss Map Entry Format
A simple map entry
.Pq mapping
takes the form:
.Bd -literal -offset indent
.Ar key Oo Fl Ns Ar mount-options Oc Ar location Ns ...
.Ed
.Pp
where
.Ar key
is the full pathname of the directory to mount when used in a direct map, or the
simple name of a subdirectory in an indirect map.
.Ar mount-options
is a comma-separated list of
.Nm mount
options, and
.Ar location
specifies a file system from which the directory may be mounted. In the case of
a simple NFS mount, the options that can be used are specified in
.Xr mount_nfs 8 ,
and
.Ar location
takes the form:
.Pp
.Dl Ar host Ns : Ns Ar pathname
.Pp
.Ar host
is the name of the host from which to mount the file system, and
.Ar pathname
is the absolute pathname of the directory to mount.
.Pp
Options to other file systems are documented in the other
.Nm mount_*
reference manual pages.
.Ss Replicated File Systems
Multiple
.Ar location
fields can be specified for replicated NFS file systems, in which case
.Nm
and the kernel will each try to use that information to increase availability.
If the read-only flag is set in the map entry,
.Nm automountd
mounts a list of locations that the kernel may use, sorted by several criteria.
Only locations available at mount time will be mounted, and thus be available to
the kernel. When a server does not respond, the kernel will switch to an
alternate server. The sort ordering of
.Nm
is used to determine how the next server is chosen. If the read-only flag is not
set,
.Nm
will mount the best single location, chosen by the same sort ordering, and new
servers will only be chosen when an unmount has been possible, and a remount is
done. Servers on the same local subnet are given the strongest preference, and
servers on the local net are given the second strongest preference. Among
servers equally far away, response times will determine the order if no
weighting factors
.Pq see below
are used.
.Pp
If the list includes server locations using both the NFS Version 2 Protocol and
the NFS Version 3 Protocol,
.Nm
will choose only a subset of the server locations on the list, so that all
entries will be the same protocol. It will choose servers with the NFS Version 3
Protocol so long as an NFS Version 2 Protocol server on a local subnet will not
be ignored. See the FIXME for additional details.
.Pp
If each
.Ar location
in the list shares the same
.Ar pathname
then a single
.Ar location
may be used with a comma-separated list of hostnames:
.Bd -literal -offset indent
.Ar hostname Ns , Ns Ar hostname Ns ...: Ns Ar pathname
.Ed
.Pp
Requests for a server may be weighted, with the weighting factor appended to
the server name as an integer in parentheses. Servers without a weighting are
assumed to have a value of zero
.Pq most likely to be selected .
Progressively higher values decrease the chance of being selected. In the
example,
.Bd -literal -offset indent
man -ro	alpha,bravo,charlie(1),delta(4):/usr/man
.Ed
.Pp
hosts
.Sy alpha
and
.Sy bravo
have the highest priority; host
.Sy delta
has the lowest.
.Pp
Server proximity takes priority in the selection process. In the example above,
if the server
.Sy delta
is on the same network segment as the client, but the others are on different
network segments, then
.Sy delta
will be selected; the weighting value is ignored. The weighting has effect only
when selecting between servers with the same network proximity. The automounter
always selects the localhost over other servers on the same network segment,
regardless of weighting.
.Pp
In cases where each server has a different export point, the weighting can
still be applied. For example:
.Bd -literal -offset indent
man	-ro	alpha:/usr/man bravo,charlie(1):/usr/share/man \e
		delta(3):/export/man
.Ed
.Pp
A mapping can be continued across input lines by escaping the NEWLINE with a
backslash
.Pq Qq Sy \e .
Comments begin with a number sign
.Pq Qq Sy #
and end at the subsequent NEWLINE.
.Ss Map Key Substitution
The ampersand
.Pq Qq Sy \*(Am
character is expanded to the value of the
.Ar key
field for the entry in which it occurs. In this case:
.Bd -literal -offset indent
jane	sparcserver:/home/&
.Ed
.Pp
the
.Sy \*(Am
expands to
.Sy jane .
.Ss Wildcard Key
The asterisk
.Pq Qq Sy *
character, when supplied as the
.Ar key
field, is recognized as the catch-all entry. Such an entry will match any key
not previously matched. For instance, if the following entry appeared in the
indirect map for
.Pa /config :
.Bd -literal -offset indent
*	&:/export/config/&
.Ed
.Pp
this would allow automatic mounts in
.Pa /config
of any remote file system whose location could be specified as:
.Bd -literal -offset indent
.Ar hostname Ns : Ns Pa /export/config/ Ns Ar hostname
.Ed
.Pp
Note that the wildcard key does not work in conjunction with the
.Fl browse
option.
.Ss Variable Substitution
Client specific variables can be used within an
.Nm
map. For instance, if
.Sy $HOST
appeared within a map,
.Nm
would expand it to its current value for the client's host name. Supported
variables are:
.Bl -column "PLATFORM" "arch -k or uname -m"
.It Sy NAME Ta Sy OUTPUT OF Ta Sy DESCRIPTION (EXAMPLE)
.It Ev ARCH
.Ta Nm arch
.Ta architecture name
.Pq Qq Sy sun4
.It Ev CPU
.Ta Nm uname Fl p
.Ta processor type
.Pq Qq Sy sparc
.It Ev HOST
.Ta Nm uname Fl n
.Ta host name
.Pq Qq Sy myhost
.It Ev KARCH
.Ta Nm arch Fl k No or Nm uname Fl m
.Ta kernel architecture name or machine hardware name
.Pq Qq Sy sun4u
.It Ev OSNAME
.Ta Nm uname Fl s
.Ta OS name
.Pq Qq Sy SunOS
.It Ev OSREL
.Ta Nm name Fl r
.Ta OS release name
.Pq Qq Sy 5.3
.It Ev OSVERS
.Ta Nm uname Fl v
.Ta OS version
.Pq Qq Sy beta1.0
.It Ev NATISA
.Ta Nm isainfo Fl n
.Ta native instruction set architecture for the system
.Pq Qq Sy sparcv9
.It Ev PLATFORM
.Ta Nm uname Fl i
.Ta platform name
.Pq Qq Sy SUNW,Sun-Fire-V240
.El
.Pp
If a reference needs to be protected from affixed characters, you can surround
the variable name with curly braces
.Pq Qq Sy \(lC Ns Sy \(rC .
.Ss Multiple Mounts
A multiple mount entry takes the form:
.Bd -literal -offset indent
.Ar key Oo Fl Ns Ar mount-options Oc Oo Oo Ar mountpoint Oc
.Oo Fl Ns Ar mount-options Oc  Ar location Ns ... Oc Ns ...
.Ed
.Pp
The initial
.Ar mountpoint
is optional for the first mount and mandatory for all subsequent mounts. The
optional
.Ar mountpoint
is taken as a pathname relative to the directory named by
.Ar key .
If
.Ar mountpoint
is omitted in the first occurrence, a
.Ar mountpoint
of
.Pa /
.Pq root
is implied.
.Pp
Given an entry in the indirect map for
.Pa /src :
.Bd -literal -offset indent
beta	-ro \e
	/		svr1,svr2:/export/src/beta  \e
	/1.0		svr1,svr2:/export/src/beta/1.0 \e
	/1.0/man	svr1,svr2:/export/src/beta/1.0/man
.Ed
.Pp
All offsets must exist on the server under
.Sy beta .
.Nm
will automatically mount
.Pa /src/beta ,
.Pa /src/beta/1.0 ,
and
.Pa /src/beta/1.0/man ,
as needed, from either
.Sy svr1
or
.Sy svr2 ,
whichever host is nearest and responds first.
.Ss Other File System Types
The automounter assumes NFS mounts as a default file system type. Other file
system types can be described using the
.Sy fstype
mount option. Other mount options specific to this file system type can be
combined with the
.Sy fstype
option. The location field must contain information specific to the file system
type. If the location field begins with a slash, a colon character must be
prepended, for instance, to mount a CD file system:
.Bd -literal -offset indent
cdrom	-fstype=hsfs,ro	:/dev/sr0
.Ed
.Pp
or to perform an
.Nm autofs
mount:
.Bd -literal -offset indent
src	-fstype=autofs	auto_src
.Ed
.Pp
Use this procedure only if you are not using Volume Manager.
.Pp
See the
.Sx NOTES
section for information on option inheritance.
.Ss Indirect Maps
An indirect map allows you to specify mappings for the subdirectories you wish
to mount under the
.Ar directory
indicated on the command line. In an indirect map, each
.Ar key
consists of a simple name that refers to one or more file systems that are to be
mounted as needed.
.Ss Direct Maps
Entries in a direct map are associated directly with
.Nm autofs
mount points. Each
.Ar key
is the full pathname of an
.Nm autofs
mount point. The direct map as a whole is not associated with any single
directory.
.Pp
Direct maps are distinguished from indirect maps by the
.Sy \-
key. For example:
.Bd -literal -offset indent
# Master map for automounter
#
+auto_master
/net	-hosts		-nosuid,nobrowse
/home	auto_home	-nobrowse
/-	auto_direct
.Ed
.Ss Included Maps
The contents of another map can be included within a map with an entry of the
form
.Bd -literal -offset indent
.No + Ns Ar mapname
.Ed
.Pp
If
.Ar mapname
begins with a slash, it is assumed to be the pathname of a local file.
Otherwise, the location of the map is determined by the policy of the name
service switch according to the entry for the automounter in
.Pa /etc/nsswitch.conf ,
such as
.Bd -literal -offset indent
automount: files nis
.Ed
.Pp
If the name service is
.Sy files ,
then the name is assumed to be that of a local file in
.Pa /etc .
If the key being searched for is not found in the included map, the search
continues with the next entry.
.Ss Special Maps
There are two special maps available:
.Sy -hosts
and
.Sy -null .
The
.Sy -hosts
map is used with the
.Pa /net
directory and assumes that the map key is the hostname of an NFS server. The
.Nm automountd
daemon dynamically constructs a map entry from the server's list of exported
file systems. References to a directory under
.Pa /net/hermes
will refer to the corresponding directory relative to
.Sy hermes
root.
.Pp
The
.Sy -null
map cancels a previous map for the directory indicated. This is most useful in
the
.Pa /etc/auto_master
for cancelling entries that would otherwise be inherited from the
.Sy +auto_master
include entry. To be effective, the
.Sy -null
entries must be inserted before the included map entry.
.Ss Executable Maps
Local maps that have the execute bit set in their file permissions will be
executed by the automounter and provided with a key to be looked up as an
argument. The executable map is expected to return the content of an
automounter map entry on its stdout or no output if the entry cannot be
determined. A direct map cannot be made executable.
.Ss Configuration and the auto_master Map
When initiated without arguments,
.Nm
consults the master map for a list of
.Nm autofs
mount points and their maps. It mounts any
.Nm autofs
mounts that are not already mounted, and unmounts
.Nm autofs
mounts that have been removed from the master map or direct map.
.Pp
The master map is assumed to be called
.Sy auto_master
and its location is determined by the name service switch policy. Normally the
master map is located initially as a local file
.Pa /etc/auto_master .
.Ss Browsing
The
.Nm automountd
daemon supports browsability of indirect maps. This allows all of the potential
mount points to be visible, whether or not they are mounted. The
.Sy -nobrowse
option can be added to any indirect
.Nm autofs
map to disable browsing. For example:
.Bd -literal -offset indent
/net	-hosts		-nosuid,nobrowse
/home	auto_home
.Ed
.Pp
In this case, any
.Ar hostname Ns s
would only be visible in
.Pa /net
after they are mounted, but all potential mount points would be visible under
.Pa /home .
The
.Sy -browse
option enables browsability of
.Nm autofs
file systems. This is the default for all indirect maps.
.Pp
The
.Sy -browse
option does not work in conjunction with the wildcard key.
.Ss Restricting Mount Maps
Options specified for a map are used as the default options for all the entries
in that map. They are ignored when map entries specify their own mount options.
.Pp
In some cases, however, it is desirable to force
.Sy nosuid , nodevices , nosetuid ,
or
.Sy noexec
for a complete mount map and its submounts. This can be done by specifying the
additional mount option,
.Sy -restrict .
.Bd -literal -offset indent
/home	auto_home	-restrict,nosuid,hard
.Ed
.Pp
The
.Sy -restrict
option forces the inheritance of all the restrictive options
.Sy nosuid , nodevices , nosetuid ,
and
.Sy noexec
as well as the restrict option itself. In this particular example, the
.Sy nosuid
and
.Sy restrict
option are inherited but the
.Sy hard
option is not. The
.Sy restrict
option also prevents the execution of
.Qq executable maps
and is enforced for auto mounts established by programs with fewer than all
privileges available in their zone.
.Sh FILES
.Bl -tag -width Ds
.It Pa /etc/auto_master
Master automount map.
.It Pa /etc/auto_home
Map to support automounted home directories.
.It Pa /etc/nsswitch.conf
Name service switch configuration file. See
.Xr nsswitch.conf 4 .
.El
.Sh EXIT STATUS
.Ex -std
.Sh SEE ALSO
.Xr isainfo 1 ,
.Xr ls 1 ,
.Xr svcs 1 ,
.Xr uname 1 ,
.Xr automountd 8 ,
.Xr mount 8 ,
.Xr mount_nfs 8 ,
.Xr svcadm 8 ,
.Xr autofs 4 ,
.Xr attributes 5 ,
.Xr nfssec 5 ,
.Xr smf 5
.Sh NOTES
.Nm autofs
mount points must not be hierarchically related.
.Nm
does not allow an
.Nm autofs
mount point to be created within another
.Nm autofs
mount.
.Pp
Since each direct map entry results in a new
.Nm autofs
mount such maps should be kept short.
.Pp
Entries in both direct and indirect maps can be modified at any time. The new
information is used when
.Nm automountd
next uses the map entry to do a mount.
.Pp
New entries added to a master map or direct map will not be useful until the
automount command is run to install them as new
.Nm autofs
mount points. New entries added to an indirect map may be used immediately.
.Pp
As of the Solaris 2.6 release, a listing
.Po see
.Xr ls 1
.Pc
of the
.Nm autofs
directory associated with an indirect map shows all potential mountable
entries. The attributes associated with the potential mountable entries are
temporary. The real file system attributes will only be shown once the file
system has been mounted.
.Pp
Default mount options can be assigned to an entire map when specified as an
optional third field in the master map. These options apply only to map entries
that have no mount options. Note that map entities with options override the
default options, as at this time, the options do not concatenate. The
concatenation feature is planned for a future release.
.Pp
When operating on a map that invokes an NFS mount, the default number of
retries for the automounter is 0, that is, a single mount attempt, with no
retries. Note that this is significantly different from the default
.Pq 10000
for the
.Xr mount_nfs 8
utility.
.Pp
The Network Information Service
.Pq NIS
was formerly known as Sun Yellow Pages
.Pq YP .
The functionality of the two remains the same.
.Pp
The
.Nm
service is managed by the service management facility,
.Xr smf 5 ,
under the service identifier:
.Bd -literal -offset indent
svc:/system/filesystem/autofs:default
.Ed
.Pp
Administrative actions on this service, such as enabling, disabling, or
requesting restart, can be performed using
.Xr svcadm 8 .
The service's status can be queried using the
.Xr svcs 1
command.


usr/src/cmd/fs.d/autofs/automount.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 *	automount.c
 *
 * Copyright (c) 2010, Oracle and/or its affiliates. All rights reserved.
 */


#include <ctype.h>
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <locale.h>
#include <stdarg.h>
#include <errno.h>
#include <string.h>
#include <dirent.h>
#include <signal.h>
#include <syslog.h>
#include <libshare.h>
#include <libscf.h>
#include <sys/param.h>
#include <sys/time.h>
#include <sys/vfs.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/mnttab.h>
#include <sys/mntent.h>
#include <sys/mount.h>
#include <sys/utsname.h>
#include <sys/tiuser.h>
#include <rpc/rpc.h>
#include <rpcsvc/nfs_prot.h>
#include <nsswitch.h>
#include <deflt.h>
#include <rpcsvc/daemon_utils.h>
#include "automount.h"
#include "smfcfg.h"

static int mkdir_r(char *);
struct autodir *dir_head;
struct autodir *dir_tail;
static struct extmnttab *find_mount();
int verbose = 0;
int trace = 0;

static void usage();
static int compare_opts(char *, char *);
static void do_unmounts();

static int mount_timeout = AUTOFS_MOUNT_TIMEOUT;

static char	*service_list[] = { AUTOMOUNTD, NULL };

/*
 * XXX
 * The following are needed because they're used in auto_subr.c and
 * we link with it. Should avoid this.
 */
mutex_t cleanup_lock;
cond_t cleanup_start_cv;
cond_t cleanup_done_cv;

int
main(int argc, char *argv[])
{
	int c;
	struct autofs_args ai;
	struct utsname utsname;
	char autofs_addr[MAXADDRLEN];
	struct autodir *dir, *d;
	struct stat stbuf;
	char *master_map = "auto_master";
	int null;
	struct extmnttab mnt, *mntp;
	struct mnttab *omntp;
	char mntopts[MAX_MNTOPT_STR];
	int mntflgs;
	int count = 0;
	char *stack[STACKSIZ];
	char **stkptr;
	char *defval;
	struct sigaction sigintact;
	int ret = 0, bufsz = 0;
	char valbuf[6];

	/*
	 * protect this command from session termination when run in background
	 * we test background by whether SIGINT is ignored
	 */
	(void) sigaction(SIGINT, NULL, &sigintact);
	if (sigintact.sa_sigaction == SIG_IGN) {
		(void) signal(SIGHUP, SIG_IGN);
		(void) setsid();
	}

	/*
	 * Read in the values from SMF first before we check
	 * commandline options so the options override the SMF values.
	 */
	bufsz = 6;
	ret = autofs_smf_get_prop("timeout", valbuf, DEFAULT_INSTANCE,
	    SCF_TYPE_INTEGER, AUTOMOUNTD, &bufsz);
	if (ret == SA_OK)
		/*
		 * Ignore errno.  In event of failure, mount_timeout is
		 * already initialized to the correct value.
		 */
		mount_timeout = strtol(valbuf, (char **)NULL, 10);

	bufsz = 6;
	ret = autofs_smf_get_prop("automount_verbose", valbuf, DEFAULT_INSTANCE,
	    SCF_TYPE_BOOLEAN, AUTOMOUNTD, &bufsz);
	if (ret == SA_OK) {
		if (strncasecmp("true", valbuf, 4) == 0)
			verbose = TRUE;
	}

	put_automountd_env();

	while ((c = getopt(argc, argv, "mM:D:f:t:v?")) != EOF) {
		switch (c) {
		case 'm':
			pr_msg("Warning: -m option not supported");
			break;
		case 'M':
			pr_msg("Warning: -M option not supported");
			break;
		case 'D':
			pr_msg("Warning: -D option not supported");
			break;
		case 'f':
			pr_msg("Error: -f option no longer supported");
			usage();
			break;
		case 't':
			if (strchr(optarg, '=')) {
				pr_msg("Error: invalid value for -t");
				usage();
			}
			mount_timeout = atoi(optarg);
			break;
		case 'v':
			verbose++;
			break;
		default:
			usage();
			break;
		}
	}

	if (optind < argc) {
		pr_msg("%s: command line mountpoints/maps "
		    "no longer supported", argv[optind]);
		usage();
	}

	current_mounts = getmntlist();
	if (current_mounts == NULL) {
		pr_msg("Couldn't establish current mounts");
		exit(1);
	}

	(void) umask(0);
	ns_setup(stack, &stkptr);

	openlog("automount", LOG_PID, LOG_DAEMON);
	(void) loadmaster_map(master_map, "", stack, &stkptr);
	if (dir_head != NULL) {
		/*
		 * automount maps found. enable services as needed.
		 */
		_check_services(service_list);
	}

	closelog();

	if (uname(&utsname) < 0) {
		pr_msg("uname: %m");
		exit(1);
	}
	(void) strcpy(autofs_addr, utsname.nodename);
	(void) strcat(autofs_addr, ".autofs");
	ai.addr.buf	= autofs_addr;
	ai.addr.len	= strlen(ai.addr.buf);
	ai.addr.maxlen	= ai.addr.len;

	ai.mount_to	= mount_timeout;
	ai.rpc_to	= AUTOFS_RPC_TIMEOUT;

	/*
	 * Mount the daemon at its mount points.
	 */
	for (dir = dir_head; dir; dir = dir->dir_next) {

		/*
		 * Skip null entries
		 */
		if (strcmp(dir->dir_map, "-null") == 0)
			continue;

		/*
		 * Skip null'ed entries
		 */
		null = 0;
		for (d = dir->dir_prev; d; d = d->dir_prev) {
			if (strcmp(dir->dir_name, d->dir_name) == 0)
				null = 1;
		}
		if (null)
			continue;

		/*
		 * Check whether there's already an entry
		 * in the mnttab for this mountpoint.
		 */
		if (mntp = find_mount(dir->dir_name, 1)) {
			/*
			 * If it's not an autofs mount - don't
			 * mount over it.
			 */
			if (strcmp(mntp->mnt_fstype, MNTTYPE_AUTOFS) != 0) {
				pr_msg("%s: already mounted",
				    mntp->mnt_mountp);
				continue;
			}

			/*
			 * Compare the mnttab entry with the master map
			 * entry.  If the map or mount options are
			 * different, then update this information
			 * with a remount.
			 */
			if (strcmp(mntp->mnt_special, dir->dir_map) == 0 &&
			    compare_opts(dir->dir_opts,
			    mntp->mnt_mntopts) == 0) {
				continue;	/* no change */
			}

			/*
			 * Check for an overlaid direct autofs mount.
			 * Cannot remount since it's inaccessible.
			 */
			omntp = (struct mnttab *)mntp;
			if (hasmntopt(omntp, "direct") != NULL) {
				mntp = find_mount(dir->dir_name, 0);
				omntp = (struct mnttab *)mntp;
				if (hasmntopt(omntp, "direct") == NULL) {
					if (verbose)
						pr_msg("%s: cannot remount",
						    dir->dir_name);
					continue;
				}
			}

			dir->dir_remount = 1;
		}

		/*
		 * Create a mount point if necessary
		 * If the path refers to an existing symbolic
		 * link, refuse to mount on it.  This avoids
		 * future problems.
		 */
		if (lstat(dir->dir_name, &stbuf) == 0) {
			if ((stbuf.st_mode & S_IFMT) != S_IFDIR) {
				pr_msg("%s: Not a directory", dir->dir_name);
				continue;
			}
		} else {
			if (mkdir_r(dir->dir_name)) {
				pr_msg("%s: %m", dir->dir_name);
				continue;
			}
		}

		ai.path 	= dir->dir_name;
		ai.opts		= dir->dir_opts;
		ai.map		= dir->dir_map;
		ai.subdir	= "";
		ai.direct 	= dir->dir_direct;
		if (dir->dir_direct)
			ai.key = dir->dir_name;
		else
			ai.key = "";

		(void) sprintf(mntopts, "ignore,%s",
		    dir->dir_direct  ? "direct" : "indirect");
		if (dir->dir_opts && *dir->dir_opts) {
			(void) strcat(mntopts, ",");
			(void) strcat(mntopts, dir->dir_opts);
		}
		mntflgs = MS_OPTIONSTR | (dir->dir_remount ? MS_REMOUNT : 0);
		if (mount(dir->dir_map, dir->dir_name, MS_DATA | mntflgs,
		    MNTTYPE_AUTOFS, &ai, sizeof (ai), mntopts,
		    MAX_MNTOPT_STR) < 0) {
			pr_msg("mount %s: %m", dir->dir_name);
			continue;
		}

		count++;

		if (verbose) {
			if (dir->dir_remount)
				pr_msg("%s remounted", dir->dir_name);
			else
				pr_msg("%s mounted", dir->dir_name);
		}
	}

	if (verbose && count == 0)
		pr_msg("no mounts");

	/*
	 * Now compare the /etc/mnttab with the master
	 * map.  Any autofs mounts in the /etc/mnttab
	 * that are not in the master map must be
	 * unmounted
	 */
	do_unmounts();

	return (0);
}

/*
 * Find a mount entry given
 * the mountpoint path.
 * Optionally return the first
 * or last entry.
 */
static struct extmnttab *
find_mount(mntpnt, first)
	char *mntpnt;
	int first;
{
	struct mntlist *mntl;
	struct extmnttab *found = NULL;

	for (mntl = current_mounts; mntl; mntl = mntl->mntl_next) {

		if (strcmp(mntpnt, mntl->mntl_mnt->mnt_mountp) == 0) {
			found = mntl->mntl_mnt;
			if (first)
				break;
		}
	}

	return (found);
}

static char *ignore_opts[] = {"ignore", "direct", "indirect", "dev", NULL};

/*
 * Compare mount options
 * ignoring "ignore", "direct", "indirect"
 * and "dev=".
 */
static int
compare_opts(opts, mntopts)
	char *opts, *mntopts;
{
	char optbuf1[MAX_MNTOPT_STR], *s = optbuf1;
	char optbuf2[MAX_MNTOPT_STR];
	char **opttbl1, **opttbl2;
	int nopts1, nopts2;
	char *ostart, *optr, *valp;
	int j, i, notsame;

	opttbl1 = opttbl2 = NULL;
	/*
	 * Parse the two option strings to split them both into
	 * lists of individual options.
	 */
	if (mntopts != NULL)
		(void) strcpy(s, mntopts);
	else
		*s = '\0';
	if (*s != '\0')
		nopts1 = 1;
	else
		nopts1 = 0;
	for (s = strchr(s, ','); s != NULL; s = strchr(s, ',')) {
		nopts1++;
		s++;
	}
	if (nopts1)
		if ((opttbl1 = memalign(sizeof (char *),
			nopts1 * sizeof (char *))) == NULL)
			return (1);
	nopts1 = 0;
	s = optbuf1;
	for (ostart = optr = s; *optr != '\0'; ostart = optr) {
		if (getsubopt(&optr, ignore_opts, &valp) == -1) {
			opttbl1[nopts1++] = ostart;
		}
	}
	s = optbuf2;
	if (opts != NULL)
		(void) strcpy(s, opts);
	else
		*s = '\0';
	if (*s != '\0')
		nopts2 = 1;
	else
		nopts2 = 0;
	for (s = strchr(s, ','); s != NULL; s = strchr(s, ',')) {
		nopts2++;
		s++;
	}
	if (nopts2)
		if ((opttbl2 = memalign(sizeof (char *),
			nopts2 * sizeof (char *))) == NULL) {
			notsame = 1;
			goto done;
		}
	nopts2 = 0;
	s = optbuf2;
	for (ostart = optr = s; *optr != '\0'; ostart = optr) {
		if (getsubopt(&optr, ignore_opts, &valp) == -1) {
			opttbl2[nopts2++] = ostart;
		}
	}
	if (nopts2 != nopts1) {
		notsame = 1;
		goto done;
	}
	notsame = 0;
	for (i = 0; i < nopts1; i++) {
		notsame = 1;
		for (j = 0; j < nopts2; j++) {
			if (strcmp(opttbl1[i], opttbl2[j]) == 0) {
				notsame = 0;
				break;
			}
		}
		if (notsame)
			break;
	}

done:
	free(opttbl1);
	free(opttbl2);
	return (notsame);
}

static void
usage()
{
	pr_msg("Usage: automount  [ -v ]  [ -t duration ]");
	exit(1);
	/* NOTREACHED */
}

/*
 * Unmount any autofs mounts that
 * aren't in the master map
 */
static void
do_unmounts()
{
	struct mntlist *mntl;
	struct extmnttab *mnt;
	struct mnttab *omnt;
	struct autodir *dir;
	int current;
	int count = 0;
	struct zone_summary *zsp;

	zsp = fs_get_zone_summaries();
	if (zsp == NULL) {
		pr_msg("Couldn't establish active zones");
		exit(1);
	}
	for (mntl = current_mounts; mntl; mntl = mntl->mntl_next) {
		mnt = mntl->mntl_mnt;
		omnt = (struct mnttab *)mnt;
		if (strcmp(mnt->mnt_fstype, MNTTYPE_AUTOFS) != 0)
			continue;
		if (fs_mount_in_other_zone(zsp, mnt->mnt_mountp))
			continue;
		/*
		 * Don't unmount autofs mounts done
		 * from the autofs mount command.
		 * How do we tell them apart ?
		 * Autofs mounts not eligible for auto-unmount
		 * have the "nest" pseudo-option.
		 */
		if (hasmntopt(omnt, "nest") != NULL)
			continue;

		current = 0;
		for (dir = dir_head; dir; dir = dir->dir_next) {
			if (strcmp(dir->dir_name, mnt->mnt_mountp) == 0) {
				current = strcmp(dir->dir_map, "-null");
				break;
			}
		}
		if (current)
			continue;


		if (umount(mnt->mnt_mountp) == 0) {
			if (verbose) {
				pr_msg("%s unmounted",
				    mnt->mnt_mountp);
			}
			count++;
		}
	}
	if (verbose && count == 0)
		pr_msg("no unmounts");
}

static int
mkdir_r(dir)
	char *dir;
{
	int err;
	char *slash;

	if (mkdir(dir, 0555) == 0 || errno == EEXIST)
		return (0);
	if (errno != ENOENT)
		return (-1);
	slash = strrchr(dir, '/');
	if (slash == NULL)
		return (-1);
	*slash = '\0';
	err = mkdir_r(dir);
	*slash++ = '/';
	if (err || !*slash)
		return (err);
	return (mkdir(dir, 0555));
}

/*
 * Print an error.
 * Works like printf (fmt string and variable args)
 * except that it will subsititute an error message
 * for a "%m" string (like syslog).
 */
/* VARARGS1 */
void
pr_msg(const char *fmt, ...)
{
	va_list ap;
	char buf[BUFSIZ], *p2;
	char *p1;
	char *nfmt;

	(void) strcpy(buf, "automount: ");
	p2 = buf + strlen(buf);

	nfmt = gettext(fmt);

	for (p1 = nfmt; *p1; p1++) {
		if (*p1 == '%' && *(p1+1) == 'm') {
			(void) strcpy(p2, strerror(errno));
			p2 += strlen(p2);
			p1++;
		} else {
			*p2++ = *p1;
		}
	}
	if (p2 > buf && *(p2-1) != '\n')
		*p2++ = '\n';
	*p2 = '\0';

	va_start(ap, fmt);
	(void) vfprintf(stderr, buf, ap);
	va_end(ap);
}
